File: browser/android/brave_relaunch_utils.cc
/* Copyright (c) 2019 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

#include "chrome/android/chrome_jni_headers/BraveRelaunchUtils_jni.h"
#include "chrome/browser/browser_process.h"
#include "chrome/browser/lifetime/application_lifetime.h"
#include "chrome/browser/profiles/profile.h"
#include "chrome/browser/profiles/profile_manager.h"
#include "components/prefs/pref_service.h"

namespace chrome {
namespace android {

void JNI_BraveRelaunchUtils_Restart(JNIEnv* env) {
  // Pref changes are batched and written to disk asynchronously. The Android
  // restart path finishes activities and then SIGKILLs the browser process
  // from a sibling process (BrowserRestartActivity) without a graceful prefs
  // flush, so a setting changed shortly before a restart can be lost if it
  // hasn't been written out yet. Flush pending pref writes synchronously here,
  // mirroring BrowserProcessImpl::EndSession() on desktop, so changes survive
  // the restart.
  if (auto* local_state = g_browser_process->local_state()) {
    local_state->CommitPendingWrite();
  }
  if (auto* profile_manager = g_browser_process->profile_manager()) {
    for (Profile* profile : profile_manager->GetLoadedProfiles()) {
      if (profile->GetPrefs()) {
        profile->GetPrefs()->CommitPendingWrite();
      }
    }
  }

  chrome::AttemptRestart();
}

}  // namespace android
}  // namespace chrome

DEFINE_JNI(BraveRelaunchUtils)

File: android/java/org/chromium/chrome/browser/BraveRelaunchUtils.java
/* Copyright (c) 2019 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

package org.chromium.chrome.browser;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;

import org.jni_zero.JNINamespace;
import org.jni_zero.NativeMethods;

import org.chromium.build.annotations.NullMarked;
import org.chromium.chrome.R;

@JNINamespace("chrome::android")
@NullMarked
public class BraveRelaunchUtils {
    public static void askForRelaunch(Context context) {
        AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(context);
        alertDialogBuilder
                .setMessage(R.string.settings_require_relaunch_notice)
                .setCancelable(true)
                .setPositiveButton(
                        R.string.settings_require_relaunch_now,
                        new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int id) {
                                BraveRelaunchUtilsJni.get().restart();
                                dialog.cancel();
                            }
                        })
                .setNegativeButton(
                        R.string.settings_require_relaunch_later,
                        new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int id) {
                                dialog.cancel();
                            }
                        });
        AlertDialog alertDialog = alertDialogBuilder.create();
        alertDialog.show();
    }

    public static void askForRelaunchCustom(Context context) {
        AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(context);
        alertDialogBuilder.setTitle(R.string.reset_brave_rewards_error_title)
                .setMessage(R.string.reset_brave_rewards_error_description)
                .setCancelable(true)
                .setPositiveButton(R.string.settings_require_relaunch_now,
                        new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int id) {
                                BraveRelaunchUtilsJni.get().restart();
                                dialog.cancel();
                            }
                        })
                .setNegativeButton(R.string.settings_require_relaunch_later,
                        new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int id) {
                                dialog.cancel();
                            }
                        });
        AlertDialog alertDialog = alertDialogBuilder.create();
        alertDialog.show();
    }

    public static void restart() {
        BraveRelaunchUtilsJni.get().restart();
    }

    @NativeMethods
    interface Natives {
        void restart();
    }
}
